### PR TITLE
Make Visible History UI more ergonomic and show inherited values

### LINKED ISSUE #4190: Make Visible History UI more intuitive and display inherited values
The VH UI should be improved by:

- using a dedicated section
- renaming the feature title
- display the inherited values
- change the check box to "Override" (or something similar)

Frame 6
Frame 5
Frame 4

### LINKED ISSUE #4210: Visible History: properly handle case where the timeline starting time is large (e.g. `log_time`)
In such case, the "Absolute" setting both needs a large value and fine tuned control (based on the timeline logged data time span). Typically, the `log_time` timeline has a large start time (e.g. "now"), and a tiny time span (logging stuff is fast). This requires proper UI treatment.

image

### PR DESCRIPTION
**Pull request by abey79**
### What

Update the Visible History UI:
- It now shows up in its own collapsible section in the Selection Panel.
- It makes it more clear that the item is either inheriting or overriding the settings.
- When inheriting, the inherited settings are displayed.
- When overriding, the editor as a more intuitive UI (fixes #4190)
- The editor also addresses a bug for timeline with large values (e.g. actual dates) but small spans, like `log_time` (fixes #4210)
- The test script now has a whole lot of timelines with all sorts of spans and offsets, for test purposes.

Side effects on the code base:
- slight tweak on the querying (latest at is used whenever visible time range boundaries resolve to the same absolute time)
- added `individual_properties` to `DataResult`, to allow UI to display/edit it
- added support to erase property overrides in `DataResult::save_override`
- added a new collapsing header widget in re_ui, to match our style

Playground RRD: [test_visible_history.rrd.zip](https://github.com/rerun-io/rerun/files/13348321/test_visible_history.rrd.zip) (same as the output of `tests/python/visible_history_playground/main.py`).

### Screenshots

New "Visible Time Range" UI in the blueprint section:

image

When Override is selected:

image

When operating on "large" time (e.g. `log_time`):

image

Same when using the Default/Inherited value:

image

<br/>
<br/>
<details><summary>Older screenshots which aged like milk, for posterity.</summary>

On (supported) space views, the choice is "Default vs. Override":

image

On groups and entities, the choice is "Inherited vs. Override":

image

On timeseries space views, the default reflects the fact that the default behaviour is, in fact, different from 2D/3D space views

image

When overriding, the text becomes "editable":

image

Same, when inherited:

image

With timelines with "large" times, an offset is extracted and explicitly displayed, while the unit in the time editor adapts to the actual span:

image

Same inherited:

image

</details>

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/4222) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4222)
- [Docs preview](https://rerun.io/preview/69df7cdce2fa9c3899b2b04aa2bf487807485c02/docs) 
- [Examples preview](https://rerun.io/preview/69df7cdce2fa9c3899b2b04aa2bf487807485c02/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

**Comment by teh-cmc** [comment]
Also the thingy goes crazy if the min temporal value matches the max temporal value for some reason:
image
image

**Comment by nikolausWest** [comment]
I think we should have "Visible Time Range" be a sub-tree to "Blueprint" rather than at the same hierarchical level while still being collapsible
Screenshot 2023-11-15 at 09 28 56

File: crates/re_viewer/src/ui/visible_history.rs [hunk 1]
```diff
@@ -0,0 +1,600 @@
+use egui::NumExt as _;
+use re_data_store::{ExtraQueryHistory, VisibleHistory, VisibleHistoryBoundary};
+use re_log_types::{EntityPath, TimeType};
+use re_space_view_spatial::{SpatialSpaceView2D, SpatialSpaceView3D};
+use re_space_view_time_series::TimeSeriesSpaceView;
+use re_types_core::ComponentName;
+use re_viewer_context::{Item, SpaceViewClassName, ViewerContext};
+use re_viewport::Viewport;
+use std::collections::HashSet;
+use std::ops::RangeInclusive;
```
Comment (teh-cmc) [inline crates/re_viewer/src/ui/visible_history.rs:10]: We generally try to maintain some order in the import chaos:
```rust
// std

// external

// re_

// crate

// super
```

Comment (teh-cmc) [reply crates/re_viewer/src/ui/visible_history.rs:10]:
```suggestion
use std::collections::HashSet;
use std::ops::RangeInclusive;

use egui::NumExt as _;

use re_data_store::{ExtraQueryHistory, VisibleHistory, VisibleHistoryBoundary};
use re_log_types::{EntityPath, TimeType};
use re_space_view_spatial::{SpatialSpaceView2D, SpatialSpaceView3D};
use re_space_view_time_series::TimeSeriesSpaceView;
use re_types_core::ComponentName;
use re_viewer_context::{Item, SpaceViewClassName, ViewerContext};
use re_viewport::Viewport;

**Review by teh-cmc — APPROVED** [review]
Most of the edge-case handling code goes way above my head but LGTM otherwise.

As I've pointed out in earlier comments there are a couple UI hiccups but in any case this is such a massive improvement overall that I wouldn't consider them blockers. Your call!

**Comment by nikolausWest** [comment]
I find the distinction here between default and inherited a bit confusing. Particularly since it seems to default to "inherited" for 2D/3D views and "default" for timeli
Screenshot 2023-11-15 at 09 42 44
Screenshot 2023-11-15 at 09 42 14
ne. 

It's also quite confusing that it says inherited but
1.  not where it's from
2. it may actually not even be inheriting anything.

One option solution be to make the choice be between default and override and then write out "Inherited from <link to where it was inherited from>" if the value comes from a parent

**Comment by nikolausWest** [comment]
Screenshot 2023-11-15 at 09 51 21

I understand why you chose this behavior but I definitely find it conceptually confusing although probably practical.  My prediction is we will make this configurable in the future. In the meantime I think this needs a more detailed description on hover.

**Comment by nikolausWest** [comment]
I think this state needs some extra explanation:
Screenshot 2023-11-15 at 09 55 00

It's not immediately obvious that what this actually means is that you will not do a range query but instead do a latest at query for the current time point

**Comment by nikolausWest** [comment]
Not for this PR but it's pretty obvious that visualizing what the time range is will make this a lot easier to understand

**Comment by abey79** [comment]
> Screenshot 2023-11-15 at 09 51 21
> I understand why you chose this behavior but I definitely find it conceptually confusing although probably practical. My prediction is we will make this configurable in the future. In the meantime I think this needs a more detailed description on hover. In the case where default means you'll do a "latest at" query instead of a range query, spelling that out is probably useful.

Oh, I definitely did _not_ make that choice! That's how the entire blueprint data model is currently implemented (ie. entity properties are saved per item _per timeline type_, as opposed to _per timeline_). The UI just reflects this state of affair. Should the blueprint data model change, I'll be glad to remove this note.

**Comment by abey79** [comment]
> I'd expect not only to be shown an offsets, but also a grayed out value indicating what the final value is with that offset applied.

I was thinking of doing something like that indeed, but it somehow overlaps with #4141. I might punt on that one in that PR, then move to the display, then we can reconsider if we need further context.

> I think we should have "Visible Time Range" be a sub-tree to "Blueprint" rather than at the same hierarchical level while still being collapsible

OK.

> Not for this PR but it's pretty obvious that visualizing what the time range is will make this a lot easier to understand

* #4141 


> I think this state needs some extra explanation: Screenshot 2023-11-15 at 09 55 00
> 
> It's not immediately obvious that what this actually means is that you will not do a range query but instead do a latest at query for the current time point

In general, "from A to B" always means "from A to B if there is anything, else whatever was logged last before A". I could add a note somewhere along the lines of "When no data exists in the interval, the data last logged before the interval is shown".

**Comment by abey79** [comment]
> I find the distinction here between default and inherited a bit confusing. Particularly since it seems to default to "inherited" for 2D/3D views and "default" for timeli Screenshot 2023-11-15 at 09 42 44 Screenshot 2023-11-15 at 09 42 14 ne.
> 
> It's also quite confusing that it says inherited but
> 
> 1. not where it's from
> 2. it may actually not even be inheriting anything.
> 
> One option solution be to make the choice be between default and override and then write out "Inherited from " if the value comes from a parent

I sense bi-directional misunderstanding here—let's have a chat about this one post-standup (or whenever).

File: crates/re_viewer/src/ui/visible_history.rs (hunk 1 above)
Comment (abey79) [reply crates/re_viewer/src/ui/visible_history.rs:10]: TIL. I thought we just let rustfmt do whatever 😅

Comment (teh-cmc) [reply crates/re_viewer/src/ui/visible_history.rs:10]: `rustfmt` will format your imports one block at a time, you get to decide what the blocks look like!

**Comment by abey79** [comment]
> Also the thingy goes crazy if the min temporal value matches the max temporal value for some reason.

Fixed.

**Comment by nikolausWest** [comment]
To fix in this or a follow up PR: initialize the override with the defaults so you don't get this jumping behavior:
https://github.com/rerun-io/rerun/assets/2624717/38f874b8-6ab0-4303-a0b6-72bac74b2915

**Comment by nikolausWest** [comment]
Bug after 
1. Set override visual history
2. Drag time cursor in the plot 

https://github.com/rerun-io/rerun/assets/2624717/96727209-2465-416f-9698-7eccfec43569

**Review by jleibs — APPROVED** [review]
This looks good. Going to land so I can handle merging with my stuff as well.

**Pull request merged**